### DISCUSSION TITLE
Private member initialization with overload of from_json?

### DISCUSSION THREAD

**Discussion by stlamoureux1** [nlohmann/json#4413] [Q&A] [answered]
Thanks in advance to all contributors. I'm trying to write overloads for `to_json` and `from_json` for my classes so that I can have ordered output and custom formatting, which I don't think I can get with `NLOHMANN_DEFINE_TYPE_INTRUSIVE`. The problem is I want to assign private member variables when going json -> class, which results in a compiler error.

For now I'm just putting the overloads in the global namespace b/c docs warn against using member methods.

Class declaration (abbreviated)
```
template <class T>
class Node 
{
public:
    Node() = default;
    Node(unsigned int id, T data);
    ...
private:
    unsigned int m_id;
    T m_data;
}
```

Overload:
```
template <class T>
void from_json(const ordered_json& j, Node<T>& node) 
{
    j.at("node").at("id").get_to(node.m_id);
    j.at("node").at("data").get_to(node.m_data);
}
```

Error msgs:
```
error: 'm_id' is a private member of 'Node<int>'
    j.at("node").at("id").get_to(node.m_id);
...
error: 'm_data' is a private member of 'Node<int>' j
    .at("node").at("data").get_to(node.m_data);
...
```

Example Call:
```
TEST(NodeTest, IOTest3) 
{
    ordered_json j;
    j["node"]["id"] = 0;
    j["node"]["data"] = 0;
    auto node = j.template get<Node<int>>();
    ASSERT_EQ(node.GetId(), 0);
    ASSERT_EQ(node.GetData(), 0);
}
```

**Comment by gregmarr** [comment, marked as answer]
In order to access private data directly, you must declare the functions as friends in the same way as the `INTRUSIVE` macros.

```
    template<typename BasicJson>
    friend void to_json(BasicJson& j, const Type& t)
    {
    }
    template<typename BasicJson>
    friend void from_json(const BasicJson& j, Type& t)
    {
    }
```

**Reply by stlamoureux1** [reply]
Thank you for such a quick response. A couple follow ups if you get the chance:
1. Can you point me to the definition of those macros? I've looked and can't find them.
2. So for this use case, they should be used as member functions? ~and my Node class would need another template parameter for BasicJson type?~ (Edit: I think I see what you're saying now -- apologies -- still learning template syntax)

**Reply by gregmarr** [reply]
1. https://github.com/nlohmann/json/blob/960b763ecd144f156d05ec61f577b04107290137/include/nlohmann/detail/macro_scope.hpp#L401
2. This doesn't change how they're used.  They're still external to the class, the friend just gives them private access.

**Reply by stlamoureux1** [reply]
Now I see. Thanks again.
